### PR TITLE
Add documentation for QuadraticProgramming.

### PR DESCRIPTION
**Pull request by Algue-Rythme**
I moved the documentation of `QuadraticProgramming` from `constrained_optimization` to its own file (and section) `quadratic_programming`: I added simple examples of usage and a table recapitulating different algorithms available.

I added a subsection on unconstrained QPs: while obvious for some, it was not straightforward for me how to solve those. I guessed that conjugate gradient was the best algorithm in this case, since after derivating the optimum we just end up with a linear system whose matrix is PSD.

I also populated `changelog.rst` with new classes; I assumed it was for the (currently unreleased) version 0.1.2.

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -21,9 +21,20 @@ Constrained
 
     jaxopt.ProjectedGradient
     jaxopt.MirrorDescent
-    jaxopt.QuadraticProgramming
     jaxopt.ScipyBoundedMinimize
 
+Quadratic Programming
```
Comment (mblondel) [inline docs/api.rst:26]:
```suggestion
Quadratic programming
```

File: docs/api.rst [hunk 2]
```diff
@@ -21,9 +21,20 @@ Constrained
...
+~~~~~~~~~~~~~~~~~~~~~
+
+.. autosummary::
+  :toctree: _autosummary
+
+    jaxopt.EqualityConstrainedQP
+    jaxopt.CvxpyQP
+    jaxopt.BoxOSQP
+    jaxopt.OSQP
+    jaxopt.QuadraticProgramming
```
Comment (mblondel) [inline docs/api.rst:36]: I guess we can remove this one from the documentation (we just keep the code until v0.3 in case someone was using it)

File: docs/quadratic_programming.rst [hunk 3]
```diff
@@ -0,0 +1,316 @@
+Quadratic programming
+=====================
+
```
Comment (mblondel) [inline docs/quadratic_programming.rst:3]: This section is concerned with minimizing quadratic functions subject to equality and/or inequality constraints, also known as quadratic programming.

File: docs/quadratic_programming.rst [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+The different specificities of the solvers are summarized in the table below.  
+The best choice will depend on the usage.  
+  
+- *jit*: the algorithm can used with jit or vmap, on GPU/TPU.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:7]:
```suggestion
- *jit*: the algorithm can be used with jit or vmap, on GPU/TPU.
```

File: docs/quadratic_programming.rst [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+     - yes
+     - +++
+     - +++
+     - +++
+     - Q, c
+
+
+This table is given as rule of thumb only; on some particular instances
+some solvers may behave unexpectedly better than others.
+In front of difficulties, we suggest to test different combinations of
```
Comment (mblondel) [inline docs/quadratic_programming.rst:64]:
```suggestion
In case of difficulties, we suggest to test different combinations of
```

File: docs/quadratic_programming.rst [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+algorithms, ``maxiter`` and ``tol`` values.  
+  
+.. warning::
+
+  Those algorithms are guaranteed to converge on **convex problems** only.
+  Hence, the matrix :math:`Q` *must* be Positive Semi Definite.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:70]:
```suggestion
  Hence, the matrix :math:`Q` *must* be positive semi-definite (PSD). 
```

File: docs/quadratic_programming.rst [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+  from jaxopt import EqualityConstrainedQP
+
+  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
+  c = jnp.array([1.0, 1.0])
+  A = jnp.array([[1.0, 1.0]])
+  b = jnp.array([1.0])
+
+  qp = EqualityConstrainedQP()
+  sol, _ = qp.run(params_obj=(Q, c), params_eq=(A, b)).params
```
Comment (mblondel) [inline docs/quadratic_programming.rst:99]: I think `, _` is not needed since you already uses `.params`

File: docs/quadratic_programming.rst [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+  from jaxopt._src.eq_qp import EqualityConstrainedQP
+
+  Q = 2 * jnp.array([[3000., 0.5], [0.5, 1]])
+  c = jnp.array([1.0, 1.0])
+  A = jnp.array([[1.0, 1.0]])
+  b = jnp.array([1.0])
+
+  qp = EqualityConstrainedQP(tol=1e-5, refine_regularization=3., refine_maxiter=50)
+  sol, _ = qp.run(params_obj=(Q, c), params_eq=(A, b)).params
```
Comment (mblondel) [inline docs/quadratic_programming.rst:120]: Same

File: docs/quadratic_programming.rst [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+  print(sol.primal)
+  print(sol.dual_eq)
+
+Ill-posed problems
+~~~~~~~~~~~~~~~~~~
+
+This solver is the fastest for well posed problems, but can behave poorly on badly scaled matrices,
+or with redundant constraints.
+
+If the solver struggles to converge, set ``refine_regularization`` and ``refine_maxiter``::
```
Comment (mblondel) [inline docs/quadratic_programming.rst:110]: Maybe mention that it uses iterative refinement and link to the wikipedia article.

File: docs/quadratic_programming.rst [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
+  c = jnp.array([1.0, 1.0])
+  A = jnp.array([[1.0, 1.0]])
+  b = jnp.array([1.0])
+  G = jnp.array([[-1.0, 0.0], [0.0, -1.0]])
+  h = jnp.array([0.0, 0.0])
+
+  qp = CvxpyWrapper()
+  sol, _ = qp.run(params_obj=(Q, c), params_eq=(A, b), params_ineq=(G, h)).params
```
Comment (mblondel) [inline docs/quadratic_programming.rst:158]: Same

File: docs/quadratic_programming.rst [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+  print(sol.primal)
+  print(sol.dual_eq)
+  print(sol.dual_ineq)
+
+Only one of ``params_eq`` and ``params_ineq`` are is mandatory.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:164]:
```suggestion
It is also possible to specify only equality constraints or only inequality constraints by setting ``params_eq`` or ``params_ineq`` to ``None``.
```

File: docs/quadratic_programming.rst [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+OSQP
+~~~~
+
+This solver is a pure Jax re-implementation of OSQP algorithm.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:169]:
```suggestion
This solver is a pure JAX re-implementation of the OSQP algorithm.
```

File: docs/quadratic_programming.rst [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+It is jittable, supports pytrees and matvecs, but the precision is usually lower than :class:`CvxpyQP`.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:170]:
```suggestion
It is jittable, supports pytrees and matvecs, but the precision is usually lower than :class:`CvxpyQP` when run in float32 precision.
```

File: docs/quadratic_programming.rst [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+
+  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
+  c = jnp.array([1.0, 1.0])
+  A = jnp.array([[1.0, 1.0]])
+  b = jnp.array([1.0])
+  G = jnp.array([[-1.0, 0.0], [0.0, -1.0]])
+  h = jnp.array([0.0, 0.0])
+
+  qp = OSQP()
+  sol, state = qp.run(params_obj=(Q, c), params_eq=(A, b), params_ineq=(G, h)).params
```
Comment (mblondel) [inline docs/quadratic_programming.rst:189]: Same

File: docs/quadratic_programming.rst [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+.. math::
+
+    \min_{x,z} \frac{1}{2} x^\top Q x + c^\top x \textrm{ subject to } A x = z, l \le z \le u
+
+.. autosummary::
+  :toctree: _autosummary
+
+    jaxopt.BoxOSQP
+
+:class:`BoxOSQP` is similar to :class:`OSQP` but accepts this input format instead.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:211]:
```suggestion
:class:`BoxOSQP` is similar to :class:`OSQP` but accepts problems in the above box-constrained format instead.
```

File: docs/quadratic_programming.rst [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+The bounds ``u`` (resp. ``l``) can be set to ``inf`` (resp. ``-inf``) if required.
+
+
```
Comment (mblondel) [inline docs/quadratic_programming.rst:214]: Remove line

File: docs/quadratic_programming.rst [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+If required the algorithm can be speed up by setting ``check_primal_dual_infeasability`` to ``False``,
+by setting ``eq_qp_preconditioner`` to ``"jacobi"`` (when possible).  
+  
+.. note::
+
+  The ``tol`` parameter is used as stopping criterion when compared to primal and dual residuals.
+  For over constrained problems, or badly scaled matrices, the residuals can be high, and the solver might
+  never converge up to the desired accuracy.
+  In this case it is better to increases ``tol`` or to reduce ``maxiter``.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:240]: Not sure about `tol`. I think the best advice is to just control `maxiter` in this case

File: docs/quadratic_programming.rst [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+
+The problem takes the form:
+
+.. math::
+
+    \min_{x} \frac{1}{2} x^\top Q x + c^\top x
+
+The optimality condition rewrites :math:`\nabla_x \frac{1}{2} x^\top Q x + c^\top x=Qx+c=0`.  
+  
+Since the matrix :math:`Q` is assumed PSD, the best algorithm to solve :math:`Qx=-c` is *conjugate gradient*.
```
Comment (mblondel) [inline docs/quadratic_programming.rst:258]: While we're at it, maybe we should write a small code snippet?

File: docs/quadratic_programming.rst [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+
+
+QuadraticProgramming (deprecated)
```
Comment (mblondel) [inline docs/quadratic_programming.rst:261]: I would remove all this since we don't want any more people to use it.

File: docs/quadratic_programming.rst [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+  never converge up to the desired accuracy.
+  In this case it is better to increases ``tol`` or to reduce ``maxiter``.  
+
+Unconstrained QPs
+-----------------
+
+.. autosummary::
+  :toctree: _autosummary
+
+    jaxopt.linear_solve.solve_cg
```
Comment (mblondel) [inline docs/quadratic_programming.rst:248]: I think it's better to not list it, as it's a tool to solve an unconstrained QP but not a QP solver per se.

**Review by mblondel — CHANGES_REQUESTED** [review]
Good job overall Louis! I left some copy-edits.

**Review by mblondel — APPROVED** [review]
